File: src/main/java/essentials/special/Vote.java
package essentials.special;

import essentials.EssentialBundle;
import essentials.EssentialPlayer;
import essentials.EssentialTimer.AutoRollback;
import essentials.Global;
import io.anuke.arc.Core;
import io.anuke.arc.Events;
import io.anuke.mindustry.Vars;
import io.anuke.mindustry.entities.type.Player;
import io.anuke.mindustry.game.EventType;
import io.anuke.mindustry.game.Team;
import io.anuke.mindustry.gen.Call;
import org.json.JSONObject;

import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.nio.file.StandardOpenOption;
import java.util.ArrayList;

import static essentials.EssentialPlayer.getData;
import static essentials.Global.bundle;
import static essentials.Global.printStackTrace;
import static io.anuke.mindustry.Vars.*;
public class Vote{
    private static Player player;
    private static Player target;
    public static boolean isvoting;
    public static ArrayList<String> list = new ArrayList<>();
    public static int require = (int) Math.ceil(0.5 * playerGroup.size());

    public void main(Player player, String type, String target){
        Vote.player = player;
        if(target != null){
            Player other = Vars.playerGroup.find(p -> p.name.equalsIgnoreCase(target));
            if(other != null){
                Vote.target = other;
            }
        }
        if(playerGroup.size() <= 3){
            bundle(player, "vote-min");
            return;
        }
        if(require == 0){
            Call.sendMessage("Vote system error! Please inform the plugin developer of this issue!");
            Global.loge("targetplayer: "+target);
            Global.loge("requestplayer: "+player.name);
            Global.loge("isvoteing: "+isvoting);
            Global.loge("arraylist: "+list.toString());
            Global.loge("require: "+require);
            return;
        }

        switch(type) {
            case "gameover":
                if(!isvoting){
                    isvoting = true;
                    for (int i = 0; i < playerGroup.size(); i++) {
                        Player others = playerGroup.all().get(i);
                        bundle(others, "vote-gameover");
                    }
                    gameover.start();
                } else {
                    bundle(player, "vote-in-processing");
                }
                break;
            case "skipwave":
                if(!isvoting){
                    isvoting = true;
                    for (int i = 0; i < playerGroup.size(); i++) {
                        Player others = playerGroup.all().get(i);
                        bundle(others, "vote-skipwave");
                    }
                    skipwave.start();
                } else {
                    bundle(player, "vote-in-processing");
                }
                break;
            case "kick":
                if(!isvoting){
                    isvoting = true;
                    for (int i = 0; i < playerGroup.size(); i++) {
                        Player others = playerGroup.all().get(i);
                        bundle(others, "vote-kick");
                    }
                    kick.start();
                } else {
                    bundle(player, "vote-in-processing");
                }
                break;
            case "rollback":
                if(!isvoting){
                    isvoting = true;
                    for (int i = 0; i < playerGroup.size(); i++) {
                        Player others = playerGroup.all().get(i);
                        bundle(others, "vote-rollback");
                    }
                    rollback.start();
                } else {
                    bundle(player, "vote-in-processing");
                }
                break;
            default:
                break;
        }
    }

    public static final Thread counting = new Thread(() -> {
        ArrayList<Thread> threads = new ArrayList<>();
        if (playerGroup != null && playerGroup.size() > 0) {
            Thread work = null;
                for (int i = 0; i < playerGroup.size(); i++) {
                    int finalI = i;
                    work = new Thread(() -> {
                        Player others = playerGroup.all().get(finalI);
                        JSONObject db1 = getData(others.uuid);
                        if (db1.get("country_code") == "KR") {
                            try {
                                Thread.sleep(10000);
                                others.sendMessage(EssentialBundle.load(true, "vote-50sec"));
                                Thread.sleep(10000);
                                others.sendMessage(EssentialBundle.load(true, "vote-40sec"));
                                Thread.sleep(10000);
                                others.sendMessage(EssentialBundle.load(true, "vote-30sec"));
                                Thread.sleep(10000);
                                others.sendMessage(EssentialBundle.load(true, "vote-20sec"));
                                Thread.sleep(10000);
                                others.sendMessage(EssentialBundle.load(true, "vote-10sec"));
                                Thread.sleep(10000);
                            } catch (Exception e) {
                                printStackTrace(e);
                            }
                        } else {
                            try {
                                Thread.sleep(10000);
                                others.sendMessage(EssentialBundle.load(false, "vote-50sec"));
                                Thread.sleep(10000);
                                others.sendMessage(EssentialBundle.load(false, "vote-40sec"));
                                Thread.sleep(10000);
                                others.sendMessage(EssentialBundle.load(false, "vote-30sec"));
                                Thread.sleep(10000);
                                others.sendMessage(EssentialBundle.load(false, "vote-20sec"));
                                Thread.sleep(10000);
                                others.sendMessage(EssentialBundle.load(false, "vote-10sec"));
                                Thread.sleep(10000);
                            } catch (Exception e) {
                                printStackTrace(e);
                            }
                        }
                    });
                    threads.add(work);
                    work.start();
                }
            try {
                Thread.sleep(60000);
            } catch (InterruptedException e) {
                assert work != null;
                for (Thread thread : threads) {
                    thread.interrupt();
                }
            }
        }
    });

    private static final Thread gameover = new Thread(() -> {
        try {
            Call.sendMessage("[green][Essentials] Require [scarlet]" + require + "[green] players.");
            counting.start();
            counting.join();
        } catch (InterruptedException ignored) {
        } finally {
            if (list.size() >= require) {
                Call.sendMessage("[green][Essentials] Gameover vote passed!");
                Events.fire(new EventType.GameOverEvent(Team.sharded));
            } else {
                Call.sendMessage("[green][Essentials] [red]Gameover vote failed.");
            }
            list.clear();
            isvoting = false;
        }
    });

    private static final Thread skipwave = new Thread(() -> {
        try {
            Call.sendMessage("[green][Essentials] Require [scarlet]" + require + "[green] players.");
            counting.start();
            counting.join();
        } catch (InterruptedException ignored) {
        } finally {
            if (list.size() >= require) {
                Call.sendMessage("[green][Essentials] Skip 10 wave vote passed!");
                for (int i = 0; i < 10; i++) {
                    logic.runWave();
                }
            } else {
                Call.sendMessage("[green][Essentials] [red]Skip 10 wave vote failed.");
            }
            list.clear();
            isvoting = false;
        }
    });

    private static final Thread kick = new Thread(() -> {
        if(target != null){
            try {
                Call.sendMessage("[green][Essentials] Require [scarlet]" + require + "[green] players.");
                counting.start();
                counting.join();
            } catch (InterruptedException ignored) {
            } finally {
                if (list.size() >= require) {
                    Call.sendMessage("[green][Essentials] Player kick vote success!");
                    EssentialPlayer.addtimeban(target.name, target.uuid, 4);
                    Global.log(target.name + " / " + target.uuid + " Player has banned due to voting. " + list.size() + "/" + require);

                    Path path = Paths.get(String.valueOf(Core.settings.getDataDirectory().child("mods/Essentials/Logs/Player.log")));
                    Path total = Paths.get(String.valueOf(Core.settings.getDataDirectory().child("mods/Essentials/Logs/Total.log")));
                    try {
                        JSONObject other = getData(target.uuid);
                        String text = other.get("name") + " / " + target.uuid + " Player has banned due to voting. " + list.size() + "/" + require + "\n";
                        byte[] result = text.getBytes();
                        Files.write(path, result, StandardOpenOption.APPEND);
                        Files.write(total, result, StandardOpenOption.APPEND);
                    } catch (IOException error) {
                        printStackTrace(error);
                    }

                    netServer.admins.banPlayer(target.uuid);
                    Call.onKick(target.con, "You're banned.");
                } else {
                    for (int i = 0; i < playerGroup.size(); i++) {
                        Player others = playerGroup.all().get(i);
                        bundle(others, "vote-failed");
                    }
                }
                list.clear();
                isvoting = false;
            }
        } else {
            bundle(player, "player-not-found");
        }
    });

    private static final Thread rollback = new Thread(() -> {
        try {
            Call.sendMessage("[green][Essentials] Require [scarlet]" + require + "[green] players.");
            counting.start();
            counting.join();
        } catch (InterruptedException ignored) {
        } finally {
            if (list.size() >= require) {
                Call.sendMessage("[green][Essentials] Map rollback passed!!");
                AutoRollback rl = new AutoRollback();
                rl.load();
            } else {
                Call.sendMessage("[green][Essentials] [red]Map rollback failed.");
            }
            list.clear();
            isvoting = false;
        }
    });
}

File: src/main/kotlin/essentials/event/feature/Vote.kt
package essentials.event.feature

import arc.Core
import arc.Events
import arc.struct.Seq
import arc.util.Time
import essentials.PluginData
import essentials.data.PlayerCore
import essentials.eof.sendMessage
import essentials.event.feature.VoteType.*
import essentials.event.feature.VoteType.Map
import essentials.internal.Bundle
import essentials.internal.Log
import essentials.internal.Log.LogType
import essentials.internal.Tool
import mindustry.Vars
import mindustry.game.EventType.GameOverEvent
import mindustry.game.Team
import mindustry.gen.Call
import mindustry.gen.Groups
import mindustry.gen.Playerc
import mindustry.net.Packets

class Vote(val player: Playerc, val type: VoteType, vararg val arg: String) {
    lateinit var target: Playerc

    var voting = false
    var require = 0
    var amount = 0
    var map : mindustry.maps.Map? = null

    val voted = Seq<String>()

    fun start() {
        val bundle = Bundle(PlayerCore[player.uuid()].locale)

        if(!voting) {
            voting = true
            if (PluginData.playerData.size < 4) {
                player.sendMessage(bundle["vote.minimal"])
            } else {
                require = if (PluginData.playerData.size > 8) 6 else 2 + if (PluginData.playerData.size > 4) 1 else 0

                Tool.sendMessageAll("vote.suggester-name", player.name())
                when (type) {
                    Kick -> {
                        this.target = arg[0] as Playerc
                        Tool.sendMessageAll("vote.kick", target.name())
                    }
                    Gameover -> Tool.sendMessageAll("vote.gameover")
                    Skipwave -> {
                        println(arg[0])
                        amount = try {
                            arg[0].toInt()
                        } catch (ignored: NumberFormatException) {
                            3
                        }
                        Tool.sendMessageAll("vote.skipwave", amount.toString())
                    }
                    Rollback -> Tool.sendMessageAll("vote.rollback")
                    Map -> {
                        var map = Vars.maps.all().find { map: mindustry.maps.Map ->
                            map.name().equals(arg[1].replace('_', ' '), ignoreCase = true) || map.name()
                                .equals(arg[1], ignoreCase = true)
                        }
                        if (map == null) {
                            map = Vars.maps.all()[arg[1].toInt()]
                            if (map == null) {
                                sendMessage(player, bundle.prefix("vote.map.not-found"))
                                return
                            }
                        }

                        Tool.sendMessageAll("vote.map", map.name())
                    }
                    None -> {}
                }
                if(voting) {
                    counting.start()
                    alert.start()
                }
            }
        } else {
            player.sendMessage(bundle["vote.in-processing"])
        }
    }

    fun interrupt(){
        alert.interrupt()
        counting.interrupt()
        voting = false
    }

    private var counting : Thread = Thread()
    private var alert : Thread = Thread()

    fun success(success: Boolean) {
        // TODO 투표 성공 메세지 bundle 추가
        if (success) {
            when (type) {
                Gameover -> {
                    Log.info("Vote gameover passed!")
                    Tool.sendMessageAll("vote.gameover.done")
                    Events.fire(GameOverEvent(Team.crux))
                }
                Skipwave -> {
                    Log.info("Vote skipwave passed!")
                    Tool.sendMessageAll("vote.skipwave.done")
                    var a = 0
                    while (a < amount) {
                        Vars.logic.runWave()
                        a++
                    }
                }
                Kick -> {
                    Log.info("Vote kick passed!")
                    PlayerCore[target.uuid()].kickcount = PlayerCore[target.uuid()].kickcount + 1
                    Tool.sendMessageAll("vote.kick.done", target.name())
                    Core.app.post { target.info.lastKicked = Time.millis() + 30 * 60 * 1000 }
                    Call.kick(target.con(), Packets.KickReason.vote)
                    Log.write(LogType.Player, "log.player.kick")
                }
                Rollback -> {
                    Log.info("Vote rollback passed!")
                    Tool.sendMessageAll("vote.rollback.done")
                    AutoRollback.load()
                }
                Map -> {
                    Log.info("Vote map passed!")
                    Tool.sendMessageAll("vote.map.done")
                    AutoRollback.load(map)
                    Tool.sendMessageAll("vote.map.done")
                }
                None -> {}
            }
        } else {
            when (type) {
                Gameover -> Tool.sendMessageAll("vote.gameover.fail")
                Skipwave -> Tool.sendMessageAll("vote.skipwave.fail")
                Kick -> Tool.sendMessageAll("vote.kick.fail", target.name())
                Rollback -> Tool.sendMessageAll("vote.rollback.fail")
                Map -> Tool.sendMessageAll("vote.map.fail")
                None -> {}
            }
        }
    }

    fun set(uuid: String) {
        voted.add(uuid)
        for (others in Groups.player) {
            val p = PlayerCore[others.uuid()]
            if (!p.error && require - voted.size != -1) {
                others.sendMessage(Bundle(p.locale).prefix("vote.current-voted", voted.size.toString(), (require - voted.size).toString()))
            }
        }
        if (voted.size >= require) {
            interrupt()
            success(voted.size >= require)
        }
    }
}